sklearn/preprocessing/_encoders.py

```diff
@@ -426,6 +488,42 @@ def fit(self, X, y=None):
             return self
         else:
             self._fit(X, handle_unknown=self.handle_unknown)
+            if self.drop is None:
+                self.drop_ = None
+            elif (isinstance(self.drop, six.string_types) and
+                    self.drop == 'first'):
+                self.drop_ = np.array([cats[0] if len(cats) > 1 else None
+                                      for cats in self.categories_],
+                                      dtype=object)
+            elif not isinstance(self.drop, six.string_types):
+                try:
+                    self.drop = np.asarray(self.drop, dtype=object)
+                    droplen = len(self.drop)
+                except (ValueError, TypeError):
+                    msg = ("Wrong input for parameter `drop`. Expected "
+                           "'first', None or array of objects, got {}")
+                    raise ValueError(msg.format(type(self.drop)))
+                if droplen != len(self.categories_):
+                    msg = ("`drop` should have length equal to the number "
+                           "of features ({}), got {}")
+                    raise ValueError(msg.format(len(self.categories_),
+                                                len(self.drop)))
+                missing_drops = [(val, i) for i, val in enumerate(self.drop)
+                                 if val not in self.categories_[i]]
+                if any(missing_drops):
+                    msg = ("The following features were supposed to be "
+                           "dropped, but were not found in the training "
+                           "data.\n{}".format(
+                               "\n".join(["Val: {}, Col: {}".format(v, c)
+                                         for v, c in missing_drops])))
+                    raise ValueError(msg)
+                self.drop_ = np.array([cat for i, cat in enumerate(self.drop)
+                                       if len(self.categories_[i]) > 1],
```

Comment:
It's a bit weird to just have entries missing (and hence unaligned to `categories_`) if there's only one category in that column. Especially so if the user has specified `drop` rather than 'first' as they have here. Looking at transform below, I find it hard to believe that this behaviour is properly tested.

Reply:
OK. I've resolved this by incorporating your suggestion from elsewhere in the thread. Now, columns with only a single category are dropped entirely from the transformed version (unless it is manually specified that no category should be dropped from the feature).